Add code examples to IO Overview

cyberixae opened the pull request:
This PR adds beginner friendly code examples from #531 to IO Overview. Some of the sequence code examples might be useful for other types with slight modifications. However, I think we may need to repeat some of the more generic concepts in several parts of the documentation since people who are reading the documentation do not know where to look for the more generic documentation. They only know that they are right now working with `IO` and wondering how to deal with problems that arise from dealing with `IO`. The fact that there are other types like `Either` which are similar and support similar operations do not help them at that point. However, as they are reading through several pages of documentation they will hopefully notice the pattern between the two.

giogonzo commented on src/IO.ts:
```diff
@@ -1,6 +1,74 @@
- * If you want to represent a synchronous computation that may fail, please see `IOEither`.
+ * @file `IO<A>` represents a synchronous computation that yields a value of type `A` and **never fails**.  If you want
+ * to represent a synchronous computation that may fail, please see `IOEither`.  IO actions are terminated by calling
+ * their `run()` method that executes the computation and returns the result.
```
Maybe a note regarding `run()`  should be called sparingly or even only once at the "end of the program" in order to maintain the referential transparency and composability advantages?

grossbart replied:
I think it would also make sense to update this to say "… represents a synchronous effectful computation", because `IO` lets us describe side-effects and non-deterministic computations.

giogonzo commented on src/IO.ts:
````diff
@@ -1,6 +1,74 @@
+ * );
+ *
+ * logger(123).run();
+ * ```
````
maybe add a reference to https://github.com/gcanti/fp-ts/blob/master/src/Console.ts ?

giogonzo commented on src/IO.ts:
````diff
@@ -1,6 +1,74 @@
+ * ```
+ *
+ * It is often convenient to combine several `IO` actions into one for sequential execution.
+ * You can use `sequence_(io, array)` to do this.
````
I would clarify that `sequence_` is helpful if you want to discard the IO results

grossbart commented on src/IO.ts:
```diff
@@ -1,6 +1,74 @@
+ * const constant: IO<number> = io.of(123);
+ * constant.run()  // returns 123
+ *
+ * const random: IO<number> = new IO(() => Math.random())
```
Similar to @giogonzo's comment on `Console.ts`, you could mention [Random](https://github.com/gcanti/fp-ts/blob/master/src/Random.ts). Maybe the mention of these two would make sense after the examples or individually just before you introduce them.

grossbart commented on src/IO.ts:
```diff
@@ -1,6 +1,74 @@
+ *   bar: io.of('example')
+ * }
+ *
+ * const computation = sequenceS(io)(computations)
```
At the moment the type `Result` is not really used, I would either add it here to show what's happening:

```ts
const computation: IO<Result> = sequenceS(io)(computations)
```

Or remove the types completely (they can be inferred) because explaining how sequence works might be better done in the `Apply` docs.

grossbart commented on src/IO.ts:
```diff
@@ -1,6 +1,74 @@
+ *
+ * const safeLog = (...args): IO<void> => new IO(() => console.log(...args));
+ *
+ * const logFoo: IO<void> = safeLog('foo');
```
If you can think of a more realistic example instead of "foo" and "bar" that would be nice (some people don't know what "foo" means). If not, that is fine.

cyberixae commented:
I updated the PR based on feedback. I replaced foo and bar with children's book style examples. Not sure if this is ideal but perhaps the examples are easier to visualise even thought these are still far from real world usage examples.

grossbart commented:
Thanks for your effort, @cyberixae!

cyberixae commented:
Merging the PR did not update the actual documentation. How does the release process work? Was I supposed to include a prebuilt document in the pull request?

grossbart commented:
Ah yes, sorry, you need to run `npm run docs` for that. I'll do this now.

grossbart commented:
It's there now: https://gcanti.github.io/fp-ts/modules/IO.ts.html
